[Clang] Implement the 'counted_by' attribute

bwendling opened the pull request:
The 'counted_by' attribute is used on flexible array members. The
argument for the attribute is the name of the field member holding the
count of elements in the flexible array. This information is used to
improve the results of the array bound sanitizer and the
'__builtin_dynamic_object_size' builtin. The 'count' field member must
be within the same non-anonymous, enclosing struct as the flexible array
member. For example:

```
  struct bar;
  struct foo {
    int count;
    struct inner {
      struct {
        int count; /* The 'count' referenced by 'counted_by' */
      };
      struct {
        /* ... */
        struct bar *array[] __attribute__((counted_by(count)));
      };
    } baz;
  };
```

This example specifies that the flexible array member 'array' has the
number of elements allocated for it in 'count':

```
  struct bar;
  struct foo {
    size_t count;
     /* ... */
    struct bar *array[] __attribute__((counted_by(count)));
  };
```

This establishes a relationship between 'array' and 'count';
specifically that 'p->array' must have *at least* 'p->count' number of
elements available. It's the user's responsibility to ensure that this
relationship is maintained throughout changes to the structure.

In the following, the allocated array erroneously has fewer elements
than what's specified by 'p->count'. This would result in an
out-of-bounds access not not being detected:

```
  struct foo *p;

  void foo_alloc(size_t count) {
    p = malloc(MAX(sizeof(struct foo),
                   offsetof(struct foo, array[0]) + count *
                       sizeof(struct bar *)));
    p->count = count + 42;
  }
```

The next example updates 'p->count', breaking the relationship
requirement that 'p->array' must have at least 'p->count' number of
elements available:

```
  void use_foo(int index, int val) {
    p->count += 42;
    p->array[index] = val; /* The sanitizer can't properly check this access */
  }
```

In this example, an update to 'p->count' maintains the relationship
requirement:

```
  void use_foo(int index, int val) {
    if (p->count == 0)
      return;
    --p->count;
    p->array[index] = val;
  }
```

bwendling committed:
[Clang] Implement the 'counted_by' attribute

The 'counted_by' attribute is used on flexible array members. The
argument for the attribute is the name of the field member holding the
count of elements in the flexible array. This information is used to
improve the results of the array bound sanitizer and the
'__builtin_dynamic_object_size' builtin. The 'count' field member must
be within the same non-anonymous, enclosing struct as the flexible array
member.

This example specifies that the flexible array member 'array' has the
number of elements allocated for it in 'count':

  struct bar;
  struct foo {
    size_t count;
     /* ... */
    struct bar *array[] __attribute__((counted_by(count)));
  };

This establishes a relationship between 'array' and 'count';
specifically that 'p->array' must have *at least* 'p->count' number of
elements available. It's the user's responsibility to ensure that this
relationship is maintained throughout changes to the structure.

In the following, the allocated array erroneously has fewer elements
than what's specified by 'p->count'. This would result in an
out-of-bounds access not not being detected:

  struct foo *p;

  void foo_alloc(size_t count) {
    p = malloc(MAX(sizeof(struct foo),
                   offsetof(struct foo, array[0]) + count *
                       sizeof(struct bar *)));
    p->count = count + 42;
  }

The next example updates 'p->count', breaking the relationship
requirement that 'p->array' must have at least 'p->count' number of
elements available:

  void use_foo(int index, int val) {
    p->count += 42;
    p->array[index] = val; /* The sanitizer can't properly check this access */
  }

In this example, an update to 'p->count' maintains the relationship
requirement:

  void use_foo(int index, int val) {
    if (p->count == 0)
      return;
    --p->count;
    p->array[index] = val;
  }

bwendling commented:
Some (hopefully) helpful comments on this admittedly very large patch.

General Comment:

I pulled back on allowing the `count` to be in a non-anonymous struct that doesn't contain the flexible array. So this is no longer allowed:

```
struct count_not_in_substruct {
  unsigned long flags;
  unsigned char count; // expected-note {{field 'count' declared here}}
  struct A {
    int dummy;
    int array[] __counted_by(count); // expected-error {{'counted_by' field 'count' isn't within the same struct as the flexible array}}
  } a;
};
```

This should align with @rapidsna's bounds checking work and get rid of the "different behaviors for a pointer and non-pointer" that we had in the previous PR (https://github.com/llvm/llvm-project/pull/73730#discussion_r1430672564).

SEMA:

The SEMA code is a bit different from the previous PR. Most notably is that I'm not using the SEMA "lookup" utility the same way. There's no clear way to prevent lookup from climbing from the current scope up through the top-level scope. This leads to false negatives, duplicate warnings, and other horrors. Also, the `Scope` for each sub-structure is deleted by the time the entire structure's parsing is completed, making checking the entire struct after parsing more difficult. And the `isAnonymousStructOrUnion` flag isn't set during SEMA.

All of these combined to make the use of the "lookup" utility very frustrating. There are probably many improvements that could be made to the lookup, but I'd prefer to do those in a follow-up patch.

CodeGen:

One observation: the `getNonTransparentContext()` (sp?) method doesn't appear to work. In particular, if called on an anonymous struct, it will return that struct and not the enclosing "non-transparent" one. I may be misunderstanding what's meant by non-transparent in this case.

There are two ways that the `counted_by` attribute are used during code generation: for sanitizer checks, and with `__builtin_dynamic_object_size()`.

* Sanitizer check: This uses @rjmccall's suggestion (https://github.com/llvm/llvm-project/pull/73730#discussion_r1427627732) of calculating a byte offset from the flexible array to the "count." It works very well (barring some issues with negative offsets being represented by an unsigned `~1U` value). The only bit of ugliness was determining the offsets of fields that aren't in the top-level of a `RecordDecl`. (I'm curious why this hasn't come up before, but I digress.) I created `getFieldOffsetInBits()` to do this. It's not quite a hack, but could probably be a bit more elegant. (For instance, there could be a `RecordLayout` method indicating if a `FieldDecl` exists in the `RecordDecl`, but that wouldn't take sub-`RecordDecls` into account, and so on, and so on...)

* `__builtin_dynamic_object_size()`: Side-effects aren't allowed with this builtin. So we can work only with pointers that have either already been emitted or that have no side-effects. The "visitor" tries to find the "base" of the expression. If it's a `DeclRefExpr`, we should be able to emit a load of that and continue. If not, we look at `LocalDeclMap` to see if the base is already available. If not, we check to see if it has side-effects and emit it if it doesn't. Assuming one of those three happen, we grab the indices to the `count` and build a series of `GEPs` to them (which seems to be how Clang generates GEPs to nested fields). Again, there didn't appear to be an existing method to get these indices. (I assume that the general case isn't as easy as when working with C structures.)

So that's basically it. I have a program on my machine that compiles most of the examples in the `test/CodeGen/attr-counted-by.c` and it works well. I don't know if there's a unittest like place that runs examples such as these, but I'd be happy to include that test there.

Thank you all for taking the time to review this!

Share and enjoy!

bwendling committed:
Use 'isAnonymousStructOrUnion' instead of testing for a 'DeclName'.

bwendling commented:
Pinging for after-holidays visibility. :-)

nickdesaulniers commented on clang/include/clang/Sema/TypoCorrection.h:
```diff
@@ -365,6 +365,10 @@ class DefaultFilterCCC final : public CorrectionCandidateCallback {
 template <class C>
 class DeclFilterCCC final : public CorrectionCandidateCallback {
 public:
+  explicit DeclFilterCCC(const IdentifierInfo *Typo = nullptr,
+                         NestedNameSpecifier *TypoNNS = nullptr)
```
I don't think you need `explicit` if the constructor accepts more than 1 argument.

nickdesaulniers commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -818,6 +819,189 @@ CodeGenFunction::evaluateOrEmitBuiltinObjectSize(const Expr *E, unsigned Type,
   return ConstantInt::get(ResType, ObjectSize, /*isSigned=*/true);
 }
 
+const FieldDecl *CodeGenFunction::FindFlexibleArrayMemberField(
+    ASTContext &Ctx, const RecordDecl *RD, uint64_t &Offset) {
+  const LangOptions::StrictFlexArraysLevelKind StrictFlexArraysLevel =
+      getLangOpts().getStrictFlexArraysLevel();
+  unsigned FieldNo = 0;
+
+  for (const Decl *D : RD->decls()) {
+    if (const auto *Field = dyn_cast<FieldDecl>(D);
+        Field && Decl::isFlexibleArrayMemberLike(
+                     Ctx, Field, Field->getType(), StrictFlexArraysLevel,
+                     /*IgnoreTemplateOrMacroSubstitution=*/true)) {
+      const ASTRecordLayout &Layout = Ctx.getASTRecordLayout(RD);
+      Offset += Layout.getFieldOffset(FieldNo);
+      return Field;
+    }
+
+    if (const auto *Record = dyn_cast<RecordDecl>(D))
+      if (const FieldDecl *Field =
+              FindFlexibleArrayMemberField(Ctx, Record, Offset)) {
+        const ASTRecordLayout &Layout = Ctx.getASTRecordLayout(RD);
+        Offset += Layout.getFieldOffset(FieldNo);
+        return Field;
+      }
+
+    if (isa<FieldDecl>(D))
+      ++FieldNo;
+  }
+
+  return nullptr;
+}
```
`FindFieldInTopLevelOrAnonymousStruct` looks somewhat similar.  Any chance to reuse code more between the two?

nickdesaulniers commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -850,6 +1034,13 @@ CodeGenFunction::emitBuiltinObjectSize(const Expr *E, unsigned Type,
+  if (IsDynamic) {
+    // Emit special code for a flexible array member with the "counted_by"
+    // attribute.
+    if (Value *V = emitFlexibleArrayMemberSize(E, Type, ResType))
+      return V;
```
Should `emitFlexibleArrayMemberSize` be made infallible (via asserts rather than early returns of `nullptr`)?

rapidsna commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -818,6 +819,189 @@ CodeGenFunction::evaluateOrEmitBuiltinObjectSize(const Expr *E, unsigned Type,
+  if (Idx) {
+    bool IdxSigned = Idx->getType()->isSignedIntegerType();
+    IdxInst = EmitAnyExprToTemp(Idx).getScalarVal();
+    IdxInst = IdxSigned ? Builder.CreateSExtOrTrunc(IdxInst, CountedByTy)
```
I'm not sure possibly truncating `Idx` to match the count type is the right thing to do. Should instead cast to the bigger type or `ssize_t`?

bwendling replied:
Is there an easy way to determine what's the larger type and cast to it? or do I need to get the bit counts and choose the larger one myself?

rapidsna replied:
Maybe we should just cast all them to size_t? This is also what `EmitBoundsCheck` and other part of code seem to do.
```
Builder.CreateIntCast(IdxInst, SizeTy, IdxSigned);
Builder.CreateIntCast(CountedByInst, SizeTy, CountedBySigned);
```

bwendling replied:
I could use `ResType` here too. It's supposed to be the expected result type of this code.

rapidsna commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -818,6 +819,189 @@ CodeGenFunction::evaluateOrEmitBuiltinObjectSize(const Expr *E, unsigned Type,
+    // We go ahead with the calculation here. If the index turns out to be
+    // negative, we'll catch it at the end.
+    CountedByInst =
+        Builder.CreateSub(CountedByInst, IdxInst, "", !IsSigned, IsSigned);
```
Should it be always signed in order to return a negative value when `Idx > count`?

bwendling replied:
I went ahead and defaulted to 'signed' after the subtraction. May God have mercy on our souls. :-)

rapidsna commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -944,22 +951,262 @@ static llvm::Value *getArrayIndexingBound(CodeGenFunction &CGF,
+  const Expr *VisitCompoundLiteralExpr(const CompoundLiteralExpr *E) {
+    return IsExpectedRecordDecl(E) ? E : nullptr;
+  }
+  const Expr *VisitCallExpr(const CallExpr *E) {
```
Do we need to handle calls?

bwendling replied:
I have a test where I made a call and just wanted to make sure we can handle it:

```
struct annotated {
  struct annotated *ann_array[10];
  unsigned long flags;
  struct {
    struct {
      int count;
      int array[] __counted_by(count);
    };
  };
};

struct annotated * noinline test1_call(struct annotated *p) {
  if (test1_count++) {
    __builtin_printf("already called\n");
    __builtin_trap();
  }
  return &p[test1_idx];
}

void noinline test1(struct annotated *p, int index) {
  test1_count = 0;
  test1_call(p)->array[index] = __builtin_dynamic_object_size(p->array, 1);
}
```

rapidsna replied:
Oh, I see. So `VisitCall` here is to handle the case for array-bounds checks.

rapidsna reviewed (commented):
I left some comments. Looking good overall but `LocalDeclMap` doesn't seem to serve the purpose for this PR. And I'm not sure the code block using it would be actually necessary. Please see my inlined review.

bwendling commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -944,22 +951,259 @@ static llvm::Value *getArrayIndexingBound(CodeGenFunction &CGF,
+      if (const auto *R = dyn_cast<RecordDecl>(RD->getDeclContext()))
+        RD = R;
+      else
+        break;
```
I changed this to:

```
  auto GetNonAnonStructOrUnion = [](const RecordDecl *RD) {
    while (RD && RD->isAnonymousStructOrUnion()) {
      const auto *R = dyn_cast<RecordDecl>(RD->getDeclContext());
      if (!R)
        break;
      RD = R;
    }
    return RD;
  };
```

I must not have pushed the change up. This is another instance of your above question about whether `R` could ever be `nullptr`. If it can't be `nullptr`, I can change this into:

```
  auto GetNonAnonStructOrUnion = [](const RecordDecl *RD) {
    while (RD->isAnonymousStructOrUnion())
      RD = cast<RecordDecl>(RD->getDeclContext());
    return RD;
  };
```

Is that an okay assumption to make here?

bwendling replied:
It looks like I can get away with a modified form of the previous version. Done.

rapidsna replied:
But in the modified form, wouldn't `RD = cast<RecordDecl>(RD->getDeclContext());` crash if it can be a non record decl as you explained with the example?

I wonder if it should just be something like:
```
  auto GetNonAnonStructOrUnion = [](const RecordDecl *RD) {
    while (RD && RD->isAnonymousStructOrUnion()) {
      const auto *R = dyn_cast<RecordDecl>(RD->getDeclContext());
      if (!R)
        return nullptr; // <- THIS
      RD = R;
    }
    return RD;
  };

rapidsna replied:
Ah I see your new commit does this already.

bwendling commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -944,22 +951,259 @@ static llvm::Value *getArrayIndexingBound(CodeGenFunction &CGF,
+    Res = EmitDeclRefLValue(DRE).getPointer(*this);
+
+    const Type *Ty = DRE->getType().getTypePtr();
+    Res = Builder.CreateAlignedLoad(ConvertType(QualType(Ty, 0)), Res,
```
Hmm...it would always complain to me whenever I did it. It works now. :-/ I changed it. :-)

bwendling commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -944,22 +951,259 @@ static llvm::Value *getArrayIndexingBound(CodeGenFunction &CGF,
+                                    getPointerAlign(), "dre.load");
+  } else {
+    Address Addr = Address::invalid();
+    auto I = LocalDeclMap.find(FAMDecl);
```
I don't have one right now. I'll remove this.

bwendling committed:
Remove use of LocalDeclMap. Also simplify some of the logic where we can rely upon assumptions.

bwendling committed:
Return the __bdos default when we can't generate the counted_by code. Mark the result as 'signed' after subtracting from it.

kees commented:
Possibly due to bug #72032 , I can get this tree to crash using the latest `array-bounds.c` test from
https://github.com/kees/kernel-tools/tree/trunk/fortify

Specifically:

```
struct anon_struct {
        unsigned long flags;
        long count;
        int array[] __counted_by(count);
};

struct composite {
        unsigned stuff;
        struct annotated inner;
};

static struct composite * noinline alloc_composite(int index)
{
        struct composite *p;

        p = malloc(sizeof(*p) + index * sizeof(*p->inner.array));
        p->inner.count = index;

        return p;
}

       struct composite *c;
       c = alloc_composite(index);
       ... actions on c->inner.array ...
```

```
3.      array-bounds.c:363:1 <Spelling=array-bounds.c:363:6>: Generating code for declaration 'counted_by_seen_by_bdos'
4.      array-bounds.c:405:2 <Spelling=array-bounds.c:23:32>: LLVM IR generation of compound statement ('{}')
...
 #4 0x0000556574d5b858 clang::CodeGen::CodeGenTBAA::getAccessInfo(clang::QualType)
 #5 0x000055657489e25d clang::CodeGen::CodeGenModule::getTBAAAccessInfo(clang::QualType)
 #6 0x00005565748a9c20 clang::CodeGen::CodeGenModule::getNaturalTypeAlignment(clang::QualType, clang::CodeGen::LValueBaseInfo*, clang::CodeGen::TBAAAccessInfo*, bool)
 #7 0x00005565749fc4a2 EmitPointerWithAlignment(clang::Expr const*, clang::CodeGen::LValueBaseInfo*, clang::CodeGen::TBAAAccessInfo*, clang::CodeGen::KnownNonNull_t, clang::CodeGen::CodeGenFunction&) CGExpr.cpp:0:0
 #8 0x00005565749f94bd clang::CodeGen::CodeGenFunction::EmitCountedByFieldExpr(clang::Expr const*, clang::FieldDecl const*, clang::FieldDecl const*)
```

bwendling committed:
Don't assume that every Expr is a pointer.

bwendling commented:
@kees Should be fixed now.

bwendling committed:
Perform proper casts to the result type instead of attempting to handle the types ourselves throughout the function.

kees commented:
Thanks! The update fixes the anon struct issue I hit. I've found one more issue, though this appears to be a miscalculation with a pathological `count` value (i.e. `count` is signed type and contains a negative value):

```
struct annotated {
    unsigned long flags;
    int count;
    int array __counted_by(count);
};

static struct annotated * noinline alloc_annotated(int index)
{
        struct annotated *p;

        p = malloc(sizeof(*p) + index * sizeof(*p->array));
        p->count = index;

        return p;
}

...
       struct annotated *a;

      c = alloc_annotated(index);
       c->count = -1;
       printf("%zu\n", __builtin_dynamic_object_size(p->array, 1));
```

This prints a wrapped calculation instead of the expected "0".

rapidsna commented:
> This prints a wrapped calculation instead of the expected "0".

Shouldn't `getDefaultBuiltinObjectSizeResult` return `-1`? Have you tried `-2` to see if it's returning the negative count value or it happens to be equal to the default value?

rapidsna commented:
> The return value of -1 means "I don't know how large this is". 

@kees Oh, I see. I did not know such the convention. Is it documented somewhere?

kees commented:
> > but the value is nonsense, so we must return 0 so that anything checking lengths will not write anything to the array.
> 
> @kees Oh, I see. I did not know such the convention but it makes sense. Is it documented somewhere?

This is new territory (having a multiplier for finding size that may be negative), so there's nothing to document it beyond FORTIFY users needing to maintain safe checks. The only safe size to return for "impossible size" is 0 in this case, otherwise a confused state (negative `count`) can lead to FORTIFY bypasses.

bwendling committed:
Return 0 if the bounds is negative. A default value indicates that the value is unknown. In the negative bounds case that isn't the case.

bwendling commented:
@kees, the issue should be fixed with the newest push.

kees commented:
> @kees, the issue should be fixed with the newest push.

Nice! This is so extremely close. It fixed all but 1 instance in the torture testing. The remaining seems to be union order sensitive. O_o This arrangement still reports non-zero for the `int count_bytes` one, unless I move it below the `signed char count_ints` counter:

```
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <malloc.h>

#define noinline                __attribute__((__noinline__))
#define __counted_by(member)    __attribute__((__counted_by__(member)))

#define DECLARE_BOUNDED_FLEX_ARRAY(COUNT_TYPE, COUNT, TYPE, NAME)       \
        struct {                                                        \
                COUNT_TYPE COUNT;                                       \
                TYPE NAME[] __counted_by(COUNT);                        \
        }

#define DECLARE_FLEX_ARRAY(TYPE, NAME)          \
        struct {                                \
                struct { } __empty_ ## NAME;    \
                TYPE NAME[];                    \
        }

struct multi {
        unsigned long flags;
        union {
                /* count member type intentionally mismatched to induce padding */
                DECLARE_BOUNDED_FLEX_ARRAY(int, count_bytes, unsigned char, bytes);
                DECLARE_BOUNDED_FLEX_ARRAY(signed char,  count_ints,  unsigned char, ints);
                DECLARE_FLEX_ARRAY(unsigned char, unsafe);
        };
};

/* Helper to hide the allocation size by using a leaf function. */
static struct multi * noinline alloc_multi_ints(int index)
{
        struct multi *p;

        p = malloc(sizeof(*p) + index * sizeof(*p->ints));
        p->count_ints = index;

        return p;
}

/* Helper to hide the allocation size by using a leaf function. */
static struct multi * noinline alloc_multi_bytes(int index)
{
        struct multi *p;

        p = malloc(sizeof(*p) + index * sizeof(*p->bytes));
        p->count_bytes = index;

        return p;
}

int main(void)
{
        struct multi *mi, *mb;
        volatile int negative = -3;
        volatile int index = 10;

        mi = alloc_multi_ints(index);
        mi->count_ints = negative;
        printf("%zu\n", __builtin_dynamic_object_size(mi->ints, 1));

        mb = alloc_multi_bytes(index);
        mb->count_bytes = negative;
        printf("%zu\n", __builtin_dynamic_object_size(mb->bytes, 1));

        return 0;
}
```

Returns:

```
253
0
```

bwendling commented:
> > @kees, the issue should be fixed with the newest push.
> 
> Nice! This is so extremely close. It fixed all but 1 instance in the torture testing. The remaining seems to be union order sensitive. O_o This arrangement still reports non-zero for the `int count_bytes` one, unless I move it below the `signed char count_ints` counter:

This is due to it being a union. There's more than one field with a `counted_by` attribute, and we just use the first one we see...and importantly it's type. This obviously isn't correct and I can fix it. However, it does bring up an issue that I *don't* know how to fix. If you have something like:

```
printf("%zu\n", __builtin_dynamic_object_size(mi, 1));
```

we don't know out of hand whether to use the `count_bytes` or `count_ints` field to get the object size. I suggest in that case we should probably throw up our hands and give up. Thoughts?

kees commented:
Accessing `mi->ints` is unambiguous (it would use the declared `count_ints`) but I'm fine to wait and fix that later. The flex array union is gloriously rare in the kernel. As for whole object, I say pick smallest from all available under bdos(x, 1) and largest for bdos(x, 0), or just say 0. Again, corner case of a corner case.

bwendling commented:
Like I said, I can fix the `mi->ints` issue. It's when we're trying to ask for the size of the whole struct with more than one `__counted_by` attribute that we don't be able to handle.

bwendling committed:
Handle unions with 'counted_by' attributes properly.

A union may have more than one entry with a 'counted_by' attribute. We
need to check the name to ensure we're getting the correct field. We
also can't calculate the total size of an object if there's a union with
multiple 'counted_by' attributes, because we won't know which field to
use for __bdos.

bwendling committed:
Regenerate

bwendling committed:
Reformat

kees commented:
Thanks! It's fixed for me now. I continue to be highly unlucky, though, and keep finding weird stuff. It seems to segfault just parsing libc headers under `-D_FORTIFY_SOURCE=3` (but not `=2` or lower):

```
0.      Program arguments: /srv/built-compilers/llvm/counted_by/install/bin/clang -Wall -O2 -fstrict-flex-arrays=3 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fsanitize=bounds -fsanitize=object-size -fsanitize-undefined-trap-on-error -c -o array-bounds.o array-bounds.c                                                                                                                        
1.      <eof> parser at end of file                                                                                             
2.      Per-file LLVM IR generation                                                                                             
3.      /usr/include/x86_64-linux-gnu/bits/string_fortified.h:57:1 <Spelling=/usr/include/x86_64-linux-gnu/bits/string_fortified.h:57:8>: Generating code for declaration 'memset'                                                                              
... 
 #4 0x000055ef8b3ac144 clang::DeclContext::decls_begin() const                                                             
 #5 0x000055ef892afcad CountCountedByAttrs(clang::RecordDecl const*) CGBuiltin.cpp:0:0                                          
```

In my header at line 57, I see:

```
__fortify_function void *
__NTH (memset (void *__dest, int __ch, size_t __len))
{
  return __builtin___memset_chk (__dest, __ch, __len,
                                 __glibc_objsize0 (__dest));
}
```

where `__glibc_objsize0` is:

```
sys/cdefs.h:# define __glibc_objsize0(__o) __builtin_dynamic_object_size (__o, 0)
```

I assume something weird is happening with `bdos` here.

bwendling commented:
@kees Could you supply a pre-processed file and command line? (It should have been emitted when you got the backtrace.)

bwendling commented:
Never mind. I think I found the issue anyway...

bwendling committed:
The RecordDecl may be null so skip if it is.

kees commented:
That's it! Everything works. :) Ship it!

kees reviewed (approved):
I can't speak to the LLVM code changes, but behaviorally, this passes all the torture tests I'd expect it to (kernel-tools/fortify/array-bounds.c and full kernel builds with its hundreds of `counted_by` annotations).

bwendling commented:
@rapidsna & @nickdesaulniers: Do you have anything to add?

rapidsna reviewed (approved):
> Do you have anything to add?

LGTM. Thanks for addressing comments!

bwendling commented:
Thanks everyone for their insightful feedback!!

dyung commented:
@bwendling the test you added `CodeGen/attr-counted-by.c` appears to be failing on quite a few bots, can you take a look?

Some failing bots:
- https://lab.llvm.org/buildbot/#/builders/164/builds/48660
- https://lab.llvm.org/buildbot/#/builders/188/builds/40261
- https://lab.llvm.org/buildbot/#/builders/245/builds/18987
- https://lab.llvm.org/buildbot/#/builders/258/builds/12107
- https://lab.llvm.org/buildbot/#/builders/249/builds/13974
- https://lab.llvm.org/buildbot/#/builders/230/builds/23480
- https://lab.llvm.org/buildbot/#/builders/124/builds/9378
- https://lab.llvm.org/buildbot/#/builders/67/builds/14012
- https://lab.llvm.org/buildbot/#/builders/98/builds/31297
- https://lab.llvm.org/buildbot/#/builders/247/builds/13055
- https://lab.llvm.org/buildbot/#/builders/139/builds/56868
- https://lab.llvm.org/buildbot/#/builders/216/builds/32771
- https://lab.llvm.org/buildbot/#/builders/275/builds/3146
- https://lab.llvm.org/buildbot/#/builders/272/builds/5208
- https://lab.llvm.org/buildbot/#/builders/271/builds/2686

nico commented:
Reverted in 2dce77201c0c6b541a53aa7a09ec06e7561e8f74 for now.

bwendling commented:
Resubmitted with fix. Sorry about the failures:

```
To https://github.com/llvm/llvm-project.git
   e8790027b169..164f85db876e  main -> main
```

seanm commented:
@bwendling is there any plan / possibility for simple expressions (with no side effects)?  Like:

```c
struct libusb_bos_dev_capability_descriptor {
	uint8_t  bLength;
	uint8_t  bDescriptorType;
	uint8_t  bDevCapabilityType;
	uint8_t  dev_capability_data[] __attribute__((counted_by(bLength - 3)));
};
```

bwendling commented:
> @bwendling is there any plan / possibility for simple expressions (with no side effects)? Like:
> 
> ```c
> struct libusb_bos_dev_capability_descriptor {
> 	uint8_t  bLength;
> 	uint8_t  bDescriptorType;
> 	uint8_t  bDevCapabilityType;
> 	uint8_t  dev_capability_data[] __attribute__((counted_by(bLength - 3)));
> };
> ```

Not right now. Apple is in the process of expanding the bounds checking code beyond checking only flexible array members. We will most likely follow their lead on `constexpr`-like expressions. But the utility of such a feature would need to be explored.

bwendling commented:
> @bwendling thanks for your reply. No idea how representative it is of other projects, but 3 of the 6 flexible arrays in [libusb](https://github.com/libusb/libusb) structures have these kinds of non-trivial counts.

Can you point them out?

> The GCC folks seem to have [considered](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=108896) adding such support, but it's not clear what, if anything, they decided.

I know about that. They haven't made a decision on it as far as I know.

seanm commented:
> Can you point them out?

Certainly. Here: https://github.com/libusb/libusb/pull/1409/commits/28394f4a570325a710ccf81959268c04b14e4dfd

rapidsna commented:
@seanm Apple's -fbounds-safety model will allow arithmetic expressions to be used for 'counted_by' attribute: https://github.com/llvm/llvm-project/blob/main/clang/docs/BoundsSafety.rst#external-bounds-annotations

We are currently working on extending the 'counted_by' attribute argument to be parsed as an expression:  https://github.com/llvm/llvm-project/pull/78000. While this patch will initially restrict the count argument to be "identification" only as it currently works (there's more work to do to actually allow an arbitrary expression), this will definitely help extend the model to support your use cases in the future.
